torch/_dynamo/variables/builder.py

```diff
@@ -1612,6 +1612,18 @@ def __eq__(self, other: object) -> bool:
 # Performs automatic dynamic dim determination.
 # Returns tuple of (dynamic_dims, constraint_dims) where each is either a list of dims or None.
 def _automatic_dynamic(e, tx, name, static_shapes):
+    if is_traceable_wrapper_subclass(e):
+        # Get dynamic dims for each tensor in the flattened list
+        dynamic_dims = []
+        constraint_dims = []
+        for tensor in get_flattened_tensors(e):
+            tensor_dynamic_dims, tensor_constraint_dims = _automatic_dynamic(
+                tensor, tx, name, static_shapes
+            )
+            dynamic_dims.append(tensor_dynamic_dims)
+            constraint_dims.append(tensor_constraint_dims)
+        return dynamic_dims, constraint_dims
```

Comment:
As discussed offline, while running the automatic dynamic algorithm recursively on the flattened components looks fine for making automatic dynamic work, I think we want this to work for the manual `mark_dynamic()` case as well.

My opinion is we should have a translation layer between outer tensor dynamic_dims -> inner tensor component dynamic dims as part of the subclass contract, possibly via a new dunder function on the subclass. Both automatic and manual dynamic dim specification would work with this. This also allows subclasses to customize how dynamic_dims are set for the inner tensors based on what makes conceptual sense for that subclass.

Maybe this is overkill, but I also want this fix for NT because I'm running into similar problems :p I'll leave it to PT2 / dynamic shapes experts to help with the proper design.

Reply:
I totally agree, I think this test is interesting as well: https://github.com/pytorch/pytorch/blob/ac4cc5dbea51a794ccd7ffcdab39a6671f625f63/test/dynamo/test_subclasses.py#L379

We are allowed too have inner_tensors with different shapes but we are currently implicitly asserting their dims are the same in terms of dynamism. 

I am not sure what to do here: https://github.com/pytorch/pytorch/blob/ac4cc5dbea51a794ccd7ffcdab39a6671f625f63/test/dynamo/test_subclasses.py#L379

@ezyang any thoughts?

Reply:
Semi-related, we also recently ran into an issue where dtensor parameters were not marked as "parameters" for cudagraphs, resulting in a bunch of extra copies.

I think we need to figure out what to do here.

Reply:
cc @soulitzer is this the same as the dynamic dim problem you were discussing a couple of weeks ago?

Reply:
Yep, for the NestedTensor case specifically, we worked around this by disabling automatic dynamic, not supporting manually setting mark_dynamic, and manually symintify/desymintify the extra dim on the outer tensor where necessary.